File: app/src/main/java/ru/atomofiron/regextool/CustomViews/NumberPreference.java
package ru.atomofiron.regextool.CustomViews;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.v7.preference.*;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

public class NumberPreference extends Preference implements NumberText.OnInputListener {

	private NumberText editText = null;
	private int initialValue = 0;

	public NumberPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
		super(context, attrs, defStyleAttr, defStyleRes);
	}

	public NumberPreference(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
	}

	public NumberPreference(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	public NumberPreference(Context context) {
		super(context);
	}

	@Override
	protected Integer onGetDefaultValue(TypedArray a, int index) {
		return a.getInt(index, 0);
	}

	@Override
	protected void onSetInitialValue(boolean restoreValue, Object defaultValue) {
		initialValue = restoreValue ? getPersistedInt(initialValue) : (Integer) defaultValue;
	}

	@Override
	public void onBindViewHolder(PreferenceViewHolder holder) {
		super.onBindViewHolder(holder);

		if (editText == null) {
			editText = new NumberText(getContext());
			editText.setFocusable(false);
			editText.setOnInputListener(this);
			editText.setLayoutParams(new LinearLayout.LayoutParams(
					LinearLayout.LayoutParams.WRAP_CONTENT,
					LinearLayout.LayoutParams.WRAP_CONTENT
			));
			ViewGroup viewGroup = holder.itemView.findViewById(android.R.id.widget_frame);
			viewGroup.addView(editText);
			viewGroup.setVisibility(View.VISIBLE);

			editText.setText(String.valueOf(initialValue));
		}
	}

	@Override
	public void onClick() {
		editText.onClick(editText);
	}

	@Override
	public void onInput(int value) {
		callChangeListener(value);
		persistInt(value);
	}
}

File: app/src/main/java/app/atomofiron/searchboxapp/custom/preference/NumberPreference.kt
package app.atomofiron.searchboxapp.custom.preference

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import app.atomofiron.searchboxapp.R
import app.atomofiron.searchboxapp.custom.view.NumberTextField

class NumberPreference(context: Context, attrs: AttributeSet) : Preference(context, attrs) {
    private var editText: NumberTextField? = null
    private var value = 0

    init {
        widgetLayoutResource = R.layout.preference_number
    }

    override fun onGetDefaultValue(array: TypedArray, index: Int): Int = array.getInt(index, 0)

    override fun onSetInitialValue(defaultValue: Any?) {
        value = getPersistedInt((defaultValue as? Int) ?: value)
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        if (editText == null) {
            val editText = holder.findViewById(R.id.number) as NumberTextField
            editText.setOnSubmitListener(::onSubmit)
            editText.setText(value.toString())
            this.editText = editText
        }
    }

    override fun onDetached() {
        super.onDetached()
        editText = null
    }

    public override fun onClick() {
        editText?.performClick()
    }

    private fun onSubmit(value: Int) {
        if (callChangeListener(value)) {
            persistInt(value)
            this.value = value
        }
    }
}